File: Array-String/CPP/two_sum.cpp
#include <vector>
#include <unordered_map>

class Solution {
public:
    // Brute Force
    vector<int> twoSum(vector<int>& nums, int target) {
        int n = nums.size();
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                if (nums[i] + nums[j] == target) {
                    return {i, j};
                }
            }
        }
        return {}; // No solution found
    }

    // using one-pass hash-map
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> umap;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            int remaining_val = target - nums[i];
            if (umap.count(remaining_val) == 1) {
                return {i, umap[remaining_val]};
            }
            umap[nums[i]] = i;
        }
        return {};
    }

    // using two-pass hash-map
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> umap;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            umap[nums[i]] = i;
        }

        for (int i = 0; i < n; i++) {
            int complement = target - nums[i];
            if (umap.count(complement) == 1 && umap[complement] != i) {
                return {i, umap[complement]};
            }
        }

        return {};
    }

};

File: Array-String/Python/two_sum.py
class Solution:
    """
    Given an array of integers nums and an integer target,
    return indices of the two numbers such that they add up to target.

    You may assume that each input would have exactly one solution,
    and you may not use the same element twice.

    You can return the answer in any order
    """
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        result_dict = dict()
        for i, value in enumerate(nums):
            remaining_value = target - nums[i]

            if remaining_value in result_dict:
                return [i, result_dict[remaining_value]]

            result_dict[value] = i